File: src/main/java/wraith/fabricaeexnihilo/api/recipes/SieveRecipe.java
package wraith.fabricaeexnihilo.api.recipes;

import net.minecraft.fluid.Fluid;
import net.minecraft.item.ItemStack;
import wraith.fabricaeexnihilo.api.crafting.FluidIngredient;
import wraith.fabricaeexnihilo.api.crafting.ItemIngredient;
import wraith.fabricaeexnihilo.api.crafting.Lootable;

import java.util.List;

public record SieveRecipe(ItemIngredient mesh,
                          FluidIngredient fluid,
                          ItemIngredient sievable,
                          List<Lootable> loot) {

    public ItemIngredient getMesh() {
        return mesh;
    }

    public FluidIngredient getFluid() {
        return fluid;
    }

    public ItemIngredient getSievable() {
        return sievable;
    }

    public List<Lootable> getLoot() {
        return loot;
    }

    public boolean test(ItemStack mesh, Fluid fluid, ItemStack sievable) {
        return this.mesh.test(mesh) && fluid != null ? this.fluid.test(fluid) : this.fluid.isEmpty() && this.sievable.test(sievable);
    }

    public boolean test(SieveRecipe other) {
        return this.mesh == other.mesh && this.fluid == other.fluid && this.sievable == other.sievable;
    }

}

File: src/main/kotlin/exnihilofabrico/api/recipes/SieveRecipe.kt
package exnihilofabrico.api.recipes

import exnihilofabrico.api.crafting.FluidIngredient
import exnihilofabrico.api.crafting.ItemIngredient
import exnihilofabrico.api.crafting.Lootable
import net.minecraft.fluid.Fluid
import net.minecraft.item.ItemStack

data class SieveRecipe(val mesh: ItemIngredient,
                       val fluid: FluidIngredient,
                       val sievable: ItemIngredient,
                       val loot: MutableCollection<Lootable>) {
    fun test(mesh: ItemStack, fluid: Fluid?, sievable: ItemStack) =
        this.mesh.test(mesh)
                && fluid?.let { this.fluid.test(fluid) } ?: this.fluid.isEmpty()
                && this.sievable.test(sievable)
    fun test(other: SieveRecipe) = this.mesh == other.mesh && this.fluid == other.fluid && this.sievable == other.sievable
}